### PR TITLE
SI-7791 Line number table reflects underlying file

### PR DESCRIPTION
**Pull request by som-snytt**
Since positions ultimately know their ultimate
positions in their ultimate source, use that line
number, ultimately, when emitting line number
table entries.

It is possible, but possibly not useful, to emit
both the actual (ultimate) line number and the
nominal one.

The `int`-valued line number of the `StackTraceElement`
is "derived" from the attribute.

note: working on clever way to test it.  Since REPL and scripter do the same thing, i.e., wrap some code, it ought to be easy to test them the same way.  The wrapping mechanism is quite different, but maybe this is an opportunity to reconsider that, under the aegis of replating.

I did see extempore's screed from 11/21/11, which you kind of wish had been backdated to 11/11/11:

```
This is a fresh start for script tests.
```

### CODE REVIEW

**Comment by retronym** [comment]
Would it suffice to have a REPL test with:

``` scala
scala> def foo = throw new Throwable; try { foo } catch { case t: Throwable => t.getStackTrace.head }
foo: Nothing
res2: StackTraceElement = .foo(<console>:7)
```

I take it the goal of this is to replace `:7` with `:1`?

**Comment by som-snytt** [comment]
That's a good question.  No, it wasn't a goal, since REPL isn't using ScriptSource or whatever that provides an underlying source.

But since you mention it, maybe it should be a goal, or another ticket.

**Comment by som-snytt** [comment]
@retronym Speaking of REPL console line numbers, looking at something else, I noticed you get double parser warnings, with different line numbers, because REPL does a quick parse before wrapping and compiling:

```
apm@mara:~/projects/snytt/src/compiler/scala/tools/nsc/ast/parser$ scalam -deprecation
Welcome to Scala version 2.11.0-M4 (OpenJDK 64-Bit Server VM, Java 1.7.0_25).
Type in expressions to have them evaluated.
Type :help for more information.

scala> val then = 7
<console>:1: warning: then is now a reserved word; usage as an identifier is deprecated
       val then = 7
           ^
<console>:7: warning: then is now a reserved word; usage as an identifier is deprecated
       val then = 7
           ^
then: Int = 7

```

**Comment by retronym** [comment]
@som-snytt Good catch. Would be good to silence the reporter during the pre-parse.

**Comment by som-snytt** [comment]
This is so hilarious I have to share it, my fix for SI-1945:

```
scala> if (false) 7
res0: AnyVal = ()

scala> else 8
res0: Int = 8
```

Nothing mind-blowingly clever, it just says if it fails to parse, try gluing it to the previous line.  Of course you have to be silent when you detect the `else 8`.

I'm trying to think of the use cases (and non-use cases), where you've got a prefix that compiles to completion and a suffix that doesn't parse until attached to it.

I haven't nuanced it yet.  And it probably won't replace paulp's leading-dot feature.

So this worked a second ago when I was gluing on typer errors too:

```
scala> List("1"->1)
res6: List[(String, Int)] = List((1,1))

scala> toMap
<console>:8: error: not found: value toMap
```

But I think just the syntax crutch is sufficient for the feature not to be confusing. (And not blow up in the face of scripts.)

**Comment by gkossakowski** [comment]
@retronym: ping on a review.

**Comment by retronym** [comment]
Still needs a test of some sort.

**Comment by som-snytt** [comment]
@retronym  @gkossakowski Sorry, I lied when I said I was working on a clever test.  I did just commit to a clever test over on SI-7805.

I thought this was going to be hard, then I thought it was easy, then it turned out to be annoying (at least while distracted).

Close to a set-up. Changed global to wrap ScriptSource in compileSources, not earlier, so DirectTest works, now I just need to capture stdout etc and strip out the first frame of the exception (to test source line).

Like icode, stack traces don't really compare across test regimes, so it might be nice if partest had a mechanism for normalizing stack traces.  Just brainstorming: strip traces from `run` output, but be able to query: origination (bottom frame), classes or packages in trace, how deep is the trace, what sources are mentioned.

**Comment by gkossakowski** [comment]
@som-snytt, @retronym: we have Bytecode test in partest which allows you to inspect arbitrary properties of bytecode by looking at ASM data structures. Would that work in this case?

Also, I'm rescheduling it for M6 so we have time to work out proper testing strategy.

**Comment by som-snytt** [comment]
@gkossakowski @retronym Yeah, it's not causing anyone to despair, so please let's push it out one M.  I'll commit the test soon.  For the record, though, it has to be pretty clever.

**Comment by som-snytt** [comment]
The test is more-or-less @retronym 's suggestion from a week ago, with some infrastructure.

It is pure coincidence that the answer happens to be "7".

File: src/compiler/scala/tools/nsc/backend/jvm/GenASM.scala [hunk 1]
```diff
@@ -2316,7 +2316,9 @@ abstract class GenASM extends SubComponent with BytecodeWriters with GenJVMASM {
               lastLineNr = currentLineNr
               val lineLab = new asm.Label
               jmethod.visitLabel(lineLab)
-              lnEntries ::= LineNumberEntry(currentLineNr, lineLab)
+              // could emit both actual and nominal line numbers, if they differ
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/backend/jvm/GenASM.scala]: Could you elaborate on this comment? Should we do this or not? If so, how?

File: src/compiler/scala/tools/nsc/Global.scala [hunk 2]
```diff
@@ -1605,12 +1606,18 @@ class Global(var currentSettings: Settings, var reporter: Reporter)
       catch { case ex: IOException => globalError(ex.getMessage()) }
     }
 
+    /** If this compilation is scripted, convert the source to a script source. */
+    def scripted(s: SourceFile) = s match {
```
Comment (retronym) [inline src/compiler/scala/tools/nsc/Global.scala]: Could this be private?

Comment (som-snytt) [inline src/compiler/scala/tools/nsc/Global.scala]: Sure can do. Privacy is a right!

File: src/compiler/scala/tools/nsc/backend/jvm/GenASM.scala (hunk 1 above)
Comment (som-snytt) [inline src/compiler/scala/tools/nsc/backend/jvm/GenASM.scala]: OK, I'll try to elaborate without belaboring (a personal weakness; I guess you'd write belabouring).

I tried it emitting both, and the stack trace seemed to pick up the bigger number, but I was too lazy to pin down the behavior. (Behaviour.)  I was going for a stack trace that would report both numbers! That would be cool.

Comment (retronym) [inline src/compiler/scala/tools/nsc/backend/jvm/GenASM.scala]: I think to emit more than one line number you need to move to [JSR 45](http://jcp.org/en/jsr/detail?id=45). That's actually sort of interesting for macros: we could emit and package a synthetic source file including macro expansions and include an alternate source map that indexes into it. But we are still some way away from that goal.

For now, I would just remove this comment.

Comment (som-snytt) [inline src/compiler/scala/tools/nsc/backend/jvm/GenASM.scala]: If I remove it, I can't belabor the point.  I guess I can live with that.

JSR 45?  I don't follow JSRs from before I graduated high school. (kidding by a large factor.)

Thanks for commenting before my next push.

**Comment by gkossakowski** [comment]
@som-snytt: is this PR ready for review?

**Comment by som-snytt** [comment]
The latest push includes @retronym hagglings and is close to his standard of excellence as scaled for ordinary mortals.  And, the test passes. So, ready for re-review.

**Comment by retronym** [comment]
LGTM.

**Pull request merged**